File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-1.java
class Solution {
    public int search(int[] nums, int target) {
    int left = 0; 
    int right = nums.length -1; 
    int mid = left + (right-left)/2; 


    while(right > left){ 
        mid = left + (right-left)/2; 
        if(nums[mid] > nums[right]){ 
            left = mid + 1; 
        } else if(nums[mid] < nums[right]){ 
            right = mid; 
        }
    }

    int pivot = left; 
    

    if(target >= nums[pivot] && target <= nums[nums.length -1]) { 
        right =  nums.length-1;
        left = pivot;
    } else { 
        right = pivot -1;
        left = 0; 
    }

     while(right > left){ 
        mid = left + (right-left)/2; 
        if(nums[mid] > target){ 
            left = mid + 1; 
        } else if(nums[mid] < left){ 
            right = mid; 
        } 
    }

    if(nums[right] != target){
        return -1; 
    } else { 
        return right;
    }
    }
}

File: Data Structures & Algorithms/time-based-key-value-store/submission-1.py
from collections import defaultdict
class TimeMap:

    def __init__(self):
        self.hashMap = defaultdict(list) 

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.hashMap[key].append((timestamp, value)) 

    def get(self, key: str, timestamp: int) -> str:
        if(self.hashMap[key][-1][0] < timestamp): 
            return ""
        
        left = 0 
        right = len(self.hashMap[key]) 
        nums = self.hashMap[key]

        while(left <= right): 
            mid = left + (right-left)//2 
          

            if(nums[mid][0] <= timestamp): 
                left = mid

            else: 
                right = mid-1

        return str(nums[left][1])


